File: library/src/main/java/com/getstream/sdk/chat/rest/request/ChannelQueryRequest.java
package com.getstream.sdk.chat.rest.request;

import com.getstream.sdk.chat.enums.Pagination;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.util.HashMap;
import java.util.Map;

public class ChannelQueryRequest extends BaseQueryChannelRequest<ChannelQueryRequest> {

    @SerializedName("messages")
    @Expose
    protected Map<String, Object> messages;

    @SerializedName("watchers")
    @Expose
    protected Map<String, Object> watchers;

    @SerializedName("members")
    @Expose
    protected Map<String, Object> members;

    @SerializedName("data")
    @Expose
    protected Map<String, Object> data;

    public ChannelQueryRequest() {
        this.watch = false;
        this.presence = false;
        this.state = true;
    }

    protected ChannelQueryRequest cloneOpts() {
        ChannelQueryRequest _this = new ChannelQueryRequest();
        _this.state = this.state;
        _this.watch = this.watch;
        _this.presence = this.presence;
        if (this.messages != null) {
            _this.messages = new HashMap<>(this.messages);
        }
        if (this.watchers != null) {
            _this.watchers = new HashMap<>(this.watchers);
        }
        if (this.members != null) {
            _this.members = new HashMap<>(this.members);
        }
        if (this.data != null) {
            _this.data = new HashMap<>(this.data);
        }
        return _this;
    }

    public ChannelQueryRequest withData(Map<String, Object> data) {
        ChannelQueryRequest clone = this.cloneOpts();
        clone.data = data;
        return clone;
    }

    public ChannelQueryRequest withPresence() {
        ChannelQueryRequest clone = this.cloneOpts();
        clone.presence = true;
        return clone;
    }

    public ChannelQueryRequest withMembers(int limit, int offset) {
        ChannelQueryRequest clone = this.cloneOpts();
        Map<String, Object> members = new HashMap<>();
        members.put("limit", limit);
        members.put("offset", offset);
        clone.members = members;
        return clone;
    }

    public ChannelQueryRequest withWatchers(int limit, int offset) {
        ChannelQueryRequest clone = this.cloneOpts();
        Map<String, Object> watchers = new HashMap<>();
        watchers.put("limit", limit);
        watchers.put("offset", offset);
        clone.watchers = watchers;
        return clone;
    }

    public ChannelQueryRequest withMessages(int limit) {
        ChannelQueryRequest clone = this.cloneOpts();
        Map<String, Object> messages = new HashMap<>();
        messages.put("limit", limit);
        clone.messages = messages;
        return clone;
    }

    public ChannelQueryRequest withMessages(Pagination direction, String messageId, int limit) {
        ChannelQueryRequest clone = this.cloneOpts();
        Map<String, Object> messages = new HashMap<>();
        messages.put("limit", limit);
        messages.put(direction.toString(), messageId);
        clone.messages = messages;
        return clone;
    }

    public Map<String, Object> getData() {
        return data;
    }
}

File: app/src/main/java/io/getstream/chat/android/core/poc/library/rest/ChannelQueryRequest.kt
package io.getstream.chat.android.core.poc.library.rest

import io.getstream.chat.android.core.poc.library.BaseQueryChannelRequest


open class ChannelQueryRequest : BaseQueryChannelRequest<ChannelQueryRequest>() {

    var messages = mutableMapOf<String, Any>()
    var watchers = mutableMapOf<String, Any>()
    var members = mutableMapOf<String, Any>()
    var data = mutableMapOf<String, Any>()

    override fun cloneOpts(): ChannelQueryRequest {
        val _this = ChannelQueryRequest()
        _this.state = state
        _this.watch = watch
        _this.presence = presence
        _this.messages = HashMap(messages)
        _this.watchers = HashMap(watchers)
        _this.members = HashMap(members)
        _this.data = HashMap(data)
        return _this
    }

    open fun withData(data: Map<String, Any>): ChannelQueryRequest {
        val clone = cloneOpts()
        clone.data = data.toMutableMap()
        return clone
    }

    override fun withPresence(): ChannelQueryRequest {
        val clone = cloneOpts()
        clone.presence = true
        return clone
    }

    fun withMembers(limit: Int, offset: Int): ChannelQueryRequest {
        val clone = cloneOpts()
        val members: MutableMap<String, Any> = HashMap()
        members["limit"] = limit
        members["offset"] = offset
        clone.members = members
        return clone
    }

    fun withWatchers(limit: Int, offset: Int): ChannelQueryRequest {
        val clone = cloneOpts()
        val watchers: MutableMap<String, Any> = HashMap()
        watchers["limit"] = limit
        watchers["offset"] = offset
        clone.watchers = watchers
        return clone
    }

    open fun withMessages(limit: Int): ChannelQueryRequest {
        val clone = cloneOpts()
        val messages: MutableMap<String, Any> = HashMap()
        messages["limit"] = limit
        clone.messages = messages
        return clone
    }

    open fun withMessages(
        direction: Pagination,
        messageId: String,
        limit: Int
    ): ChannelQueryRequest {
        val clone = cloneOpts()
        val messages: MutableMap<String, Any> = HashMap()
        messages["limit"] = limit
        messages[direction.toString()] = messageId
        clone.messages = messages
        return clone
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ChannelQueryRequest

        if (messages != other.messages) return false
        if (watchers != other.watchers) return false
        if (members != other.members) return false
        if (data != other.data) return false

        return true
    }

    override fun hashCode(): Int {
        var result = messages.hashCode()
        result = 31 * result + watchers.hashCode()
        result = 31 * result + members.hashCode()
        result = 31 * result + data.hashCode()
        return result
    }

    init {
        watch = false
        presence = false
        state = true
    }


}
